src/sentry/utils/eventuser.py

```diff
@@ -3,43 +3,152 @@
+from sentry.models.project import Project
 from sentry.snuba.dataset import Dataset, EntityKey
 from sentry.utils.avatar import get_gravatar_url
+from sentry.utils.datastructures import BidirectionalMapping
 from sentry.utils.snuba import raw_snql_query
 
 logger = logging.getLogger(__name__)
 
 REFERRER = "sentry.utils.eventuser"
 
+SNUBA_KEYWORD_MAP = BidirectionalMapping(
+    {
+        ("user_id"): "id",
+        ("user_name"): "username",
+        ("email"): "email",
+        ("ip_address_4", "ip_address_6"): "ip",
+    }
+)
+
+# The order of these keys are significant to also indicate priority
+# when used in hashing and determining uniqueness. If you change the order
+# you will break stuff.
+KEYWORD_MAP = BidirectionalMapping(
+    {
+        "user_ident": "id",
+        "username": "username",
+        "email": "email",
+        "ip_address": "ip",
+    }
+)
+
 
 @dataclass
 class EventUser:
     project_id: Optional[int]
-    email: str
-    username: str
-    name: str
-    ip_address: str
-    id: Optional[int] = None
+    email: Optional[str]
+    username: Optional[str]
+    name: Optional[str]
+    ip_address: Optional[str]
+    user_ident: Optional[int]
+    id: Optional[int] = None  # EventUser model id
 
     @staticmethod
-    def from_event(event: Event):
+    def from_event(event: Event) -> EventUser:
         return EventUser(
             id=None,
             project_id=event.project_id if event else None,
             email=event.data.get("user", {}).get("email") if event else None,
             username=event.data.get("user", {}).get("username") if event else None,
             name=event.data.get("user", {}).get("name") if event else None,
             ip_address=event.data.get("user", {}).get("ip_address") if event else None,
+            user_ident=event.data.get("user", {}).get("id") if event else None,
         )
 
     def get_display_name(self):
         return self.name or self.email or self.username
 
+    @classmethod
+    def for_projects(
+        self, projects: List[Project], keyword_filters: Mapping[str, Any]
+    ) -> List[EventUser]:
+        """
+        Fetch the EventUser with a Snuba query that exists within a list of projects
+        and valid `keyword_filters`. The `keyword_filter` keys are in `KEYWORD_MAP`.
+        """
+        oldest_project = min(projects, key=lambda item: item.date_added)
+
+        where_conditions = [
+            Condition(Column("project_id"), Op.IN, [p.id for p in projects]),
+            Condition(Column("timestamp"), Op.LT, datetime.now()),
+            Condition(Column("timestamp"), Op.GTE, oldest_project.date_added),
+        ]
+
+        for keyword, value in keyword_filters.items():
+            snuba_column = SNUBA_KEYWORD_MAP.get_key(keyword)
+            if isinstance(snuba_column, tuple):
+                for column in snuba_column:
+                    where_conditions.append(Condition(Column(column), Op.EQ, value))
```

Comment:
Now that I'm reading this I'm not sure it's correct. Won't each of these conditions be an `and`? So if we're doing an ip lookup here, it'd look for the same value in the ip4 and ip6 col? I think it should be an or.  Either way, probably worth validating with a test.

Reply:
might also need to process the ip addresses as well -- in my pr for migrating project users endpoint from eventuser, i ran into issues with testing when filtering for IP addresses since they weren't stored as strings in snuba, and ended up needing to use the IPv4StringToNum and IPv6StringToNum functions ([relevant lines](https://github.com/getsentry/sentry/pull/59497/files#diff-39e40296c1e50fd67fc0c9dd5a6f8434b7efda06de8b4963668145782f350b8aR61-R86))